src/dmd/dmangle.d

```diff
@@ -229,7 +231,22 @@ public:
     bool backrefType(Type t)
     {
         if (!t.isTypeBasic())
+        {
+            /**
+             * https://issues.dlang.org/show_bug.cgi?id=21591
+             *
+             * Special case for unmerged TypeFunctions: use the generic merged
+             * function type as backref cache key to avoid missed backrefs.
+             *
+             * Merging is based on mangling, so we need to avoid an infinite
+             * recursion by excluding the case where `t` is the function type
+             * passed to `mangleToBuffer()`: `rootType`
+             */
+            if (t != rootType && t.isTypeFunction())
+                t = t.merge2();
```

Comment:
Pinging @rainers. I've also tried `if (t.deco && t.isTypeFunction())` which worked in my case too and wouldn't require this ugly `rootType` field; I'm not sure whether it would be reliable though...

Reply:
Is it actually possible to have a function type refer to itself? How would you declare that? Or is the `rootType` test just triggering for erroneous type declarations?

Reply:
I don't think it's possible; it's just there to prevent an infinite recursion in case the `deco` hasn't been computed yet and `merge2()` calls `mangleToBuffer(tf)`, where we end up in here again.